Use full path as data of documents in the xapian index for compatibility of new zim files with older libzim

maneeshpm opened the pull request:
Fixes #523 

There is a compatibility issue when reading new zim files with older libzim used by some kiwixes. Older libzim still uses the convention where the namespace is not hidden from the path. Hence it fails to load items from the paths stored in new zim file indexes. The solution is to index the `fullPath` with namespace as the data of documents in the xapian index. This will make older libzim happy. We already have a compatibility layer in newer libzim that handles the old path convention with namespace, we just modify search iterator `get_path`(formerly `get_url`) to do the same.

The changes included in this pr are:
- Index `fullPath` in xapian title index.
- Index `fullPath` in xapian ft index.
- Introduce a new database metadata `data` that signifies what is contained as "data" of documents in the index.
- Rename `search_iterator:;get_url` to `search_iterator::get_path` and handle new namespace scheme
- Add unit test to ensure proper indexing of data

mgautierfr reviewed (changes requested):
> There is a compatibility issue when reading new zim files with older libzim used by some kiwixes. Older libzim still uses the convention where the namespace is not hidden from the path. Hence it fails to load items from the paths stored in new zim file indexes. The solution is to index the URL (full path with namespace) as the data of documents in the xapian index. This will make older libzim happy.

Mostly agree.
However, I disagree with the name `url`. Qualifying the path with a namespace doesn't create a url.
I was already disliking this name with the previous api and I use the api change as a occasion to remove it. I haven't removed it because it doesn't have a namespace. I've removed it because it is a wrong name.

**If** we reintroduce a "path with namespace" I would rather name it `fullPath`, `qualifiedPath` or `nsPath`, ...

> We already have a compatibility layer in newer libzim that handles the old path convention with namespace.

Yes, but it is not enough.
The API MUST be coherent.
The new API hide the namespace. We must hide it all the time, whatever we store in the xapian database.
So we must change https://github.com/openzim/libzim/blob/master/src/search_iterator.cpp#L100-L105 (at least) to remove the potential namespace stored as document's data.

We must add a metadata in the xapian database when we create it (https://github.com/openzim/libzim/blob/master/src/writer/xapianIndexer.cpp#L85-L98).
Same as "kind", but something about the fact we store a qualifiedPath or not.
I suggest to have a "data" key storing what is stored as documentData.
For now we would have the only one value `fullPath` (or `qualifiedPath` or `nsPath` ...).
It will allow us to change the data in the future.
If in the future we decide to remove the namespace (we will), we will change the "data" value from `fullPath` to `path`.
(We may also in another future store more things in the data, this metadata would allow use to know what is stored).

At reading level, we must read the `data` metadata (if not present, assume it is `fullPath`).
If it is `fullPath`, we need to remove the namespace everywhere necessary before returning the path to the user.
If it is `path`, we don't change the path (we need to do it now to start distributing libraries able to read "new" xapian database when we will do the change).

This way, we would have:
- A coherent API (no namespace, everywhere).
- New zims working with old library.
- New library able to read the future zims.

> - Introduce a function getURL() in dirents to get the full path in xapian handlers.
> - Introduce a variable m_url in items to store the full path of items.

Independently of the "url" name I dislike, I think we don't need it.
It would be better to extend the arguments of the indexer methods to pass the namespace and let the indexer prepend the namespace to the path
just before adding it to the document's data.
And, as we always index entries in the `C` namespace, we don't have to change the methods signature at all. Just prepend `C/` to the path.

maneeshpm commented:
Yes, having a metadata indicating the type of data stored is a good idea. Right now, a path with namespace can make it into our API only via the `Xapian::Document::get_data` method. In my opinion, if we use an extended `Xapian::Document` class, we can override the `get_data` method to release only the "filtered" data into our API, as per our current need and modifiable for the future as well. If this is done, all that is left to do would be to simply rename "URL"s to "Paths" and removing namespaces wherever necessary rather than modifying functions everywhere to use a substr of path conditionally.

As far as this PR is concerned, it seems relevant to keep only the indexing full path change and adding a metadata `data`. The rest of hiding the namespace stuff should be done in a separate pr. What do you think @mgautierfr?

mgautierfr commented:
>_Should we split this function into Entry::getPath and Entry::getFullPath and then handle tests accordingly? Or should we return getUrl() in all cases?

No

> PS: Since zim files with old namespace scheme do not have the new metadata, I have changed the search_iterator::get_path to return fullPath with namespace when the metadata is not found. This maintains consistency with the behavior of Item::getPath and Entry::getPath.

The `data` metadata describe what is in the database.
This is [`hasNewNamespaceScheme()`](https://github.com/openzim/libzim/blob/master/src/fileimpl.h#L91) which describe how we must return the path to the user. If it is true, we must not have a namespace (and so remove it if `data` is `fullPath`). If it false we must have it (and add it if needed, even if we should not have this configuration).

maneeshpm commented:
@mgautierfr I have fixed `get_path` to use `hasNewNamespaceScheme`. The 3rd and 4th commit message needs to be edited as per our new modifications. I will do that along with rebasing after the your review.

mgautierfr commented on src/search_iterator.cpp:
```diff
@@ -103,11 +104,10 @@ std::string search_iterator::get_path() const {
-    // should return full path for consistency with Item::getPath and Entry::getPath
-    if(!dbData.empty() && dbData == "fullPath") {
+    // If the archive has new namespace scheme, we return the path without namespace
+    if(impl->hasNewNamespaceScheme()) {
```
We need to remove the namespace if `hasNewNamespaceScheme()` is true AND if there is a namespace to remove (`dbData == "fullPath"`).

In the future we will generate index without namespace in it. We want the "current" libzim being ready to handle this.

maneeshpm commented:
@mgautierfr Made the necessary changes, cleared conflicts, and rebased. Even though I have added unit tests for the changes, not sure why the CI is still not happy.

mgautierfr commented on include/zim/search_iterator.h:
```diff
@@ -46,7 +46,8 @@ class search_iterator : public std::iterator<std::bidirectional_iterator_tag, En
 
-        std::string get_url() const;
+        std::string get_path() const;
+        std::string get_dbData() const;
```
Make this private, we don't want to allow a user to get the dbData.

As we are using it in test we need to have it.
But you can do as for `Archive` for some methods (https://github.com/openzim/libzim/blob/master/include/zim/archive.h#L401-L405)

maneeshpm replied:
This is a very neat solution! :+1:

mgautierfr commented on src/search_iterator.cpp:
```diff
@@ -97,10 +98,28 @@ search_iterator search_iterator::operator--(int) {
+
+    std::string path = internal->get_document().get_data();
+    std::string dbDataType = internal->search->internal->database.get_metadata("data");
+    auto impl = internal->search->m_archives.at(get_fileIndex()).getImpl();
```
The `Archive` already has a method to check if it is a new namespace scheme.
We can use it instead of getting the implementation.

(We may want to move all the internal machinery of search to use impl, but it is another change)

kelson42 commented:
@rgaudin Maybe you have a clue why the CI fails, this seems to be macOS/Homebrew specific?

rgaudin commented:
@kelson42 yes, homebrew moved to gh hosting following bintray's change. You have to run brew update in the CI before installing

mgautierfr commented:
We are good with the final code.

9 commits seems a bit too much for this code change. Maybe you can squash few of them @maneeshpm (if not all) ?

maneeshpm commented:
@mgautierfr Is this structure ok?

mgautierfr commented on test/search.cpp:
```diff
@@ -39,4 +75,30 @@ TEST(Search, searchByTitle)
+  search.set_verbose(true);
+
+  // URL should have the namespace prefix "C/"
+  ASSERT_EQ(search.begin().get_url().substr(0, 2), "te");
```
We the comment says it must start with "C/" and you test with "te" ?

It is not a problem that new added tests are failing.
In fact, this is a good thing at it ensure that the following correction actually fix a tested feature.

mgautierfr commented on src/writer/xapianIndexer.cpp:
```diff
@@ -89,10 +94,12 @@ void XapianIndexer::indexingPrelude()
     case IndexingMode::FULL:
       writableDatabase.set_metadata("valuesmap", "title:0;wordcount:1;geo.position:2");
       writableDatabase.set_metadata("kind", "fulltext");
+      writableDatabase.set_metadata("data", "fullPath");
```
This commit (and all other commit "reading" the metadata and the dbData) can be merge with the commit adding the "C/" prefix.

mgautierfr commented on include/zim/search_iterator.h:
```diff
@@ -56,6 +56,10 @@ class search_iterator : public std::iterator<std::bidirectional_iterator_tag, En
         pointer operator->() const;
 
+#ifdef ZIM_PRIVATE
+        std::string get_dbData() const;
```
This commit however can be keep separated as it introduces a somehow autonomous method.

mgautierfr reviewed (changes requested):
I think we can have 3 commits :
- Add the tests
- Add helper `get_dbData`
- Fix the code.

maneeshpm commented:
@mgautierfr Just a small change from your suggestion. I am placing the commit for `get_dbData()` first because we need to use the function for tests in the following commits.

kelson42 commented:
@mgautierfr Can we finally merge this one?

mgautierfr commented on src/search_iterator.cpp:
```diff
@@ -97,10 +99,28 @@ search_iterator search_iterator::operator--(int) {
+    }
+
+    std::string path = internal->get_document().get_data();
+    std::string dbDataType = internal->search->internal->database.get_metadata("data");
```
If `data` is empty, we must default to "fullPath".

We should probably never have a `hasNewNamespaceScheme` zim without a "data" in the xapian db, but it is better to properly handle this anyway.

maneeshpm replied:
@mgautierfr  Do we need to check this explicitly? Currently the namespace is trimmed(inside if block) only if `hasNewNamespace` is true and `dbDataType` is equal to "fullPath". In all other cases we will default to untrimmed data.
So if data is empty, we are anyway returning the untrimmed path

mgautierfr replied:
If we are dealing with zim file created only with libzim.
I don't know any other writer, but it is not a big deal to check if data is empty and set it to "fullPath".

maneeshpm replied:
aah yes, understood. Will do that :+1:

mgautierfr reviewed (changes requested):
On last change. I've missed it in previous review.

kelson42 commented:
@maneeshpm I have rebased to include the macOS CI fix

kelson42 commented:
@maneeshpm One remark on my side, commit messages should not be too long (<50 characters). Otherwise the git log is difficult to read. If you have more to say (often the case), then put it in the commit description.

maneeshpm commented:
Thank you for the rebase and the remark @kelson42. Pushing a better commit message.

maneeshpm commented:
Just a small fixup to make codecov happy. Adding a to ensure get_path works correctly.
